### PR TITLE
Make the default HttpClientFactory resilient to port exhaustion and stale DNS

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Platforms/netcore/NetCoreHttpClientFactory.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Net;
+using System.Net.Http;
+using Microsoft.Identity.Client.Http;
+
+namespace Microsoft.Identity.Client.Platforms.netcore
+{
+    internal class NetCoreHttpClientFactory : IMsalHttpClientFactory
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Platforms/netcore/NetCoreHttpClientFactory.cs:10]: This and the `NetDesktopHttpClientFactory` are the most important changes.

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/SimpleHttpClientFactory.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    /// <summary>
+    /// A simple implementation of the HttpClient factory that uses a managed HttpClientHandler
+    /// </summary>
+    /// <remarks>
+    /// This implementation is not suitable for high-scale applications / confidential client scenarios
+    /// because creating new HttpClients leads to a port exhaustion issue. 
+    /// Mobile platforms should use HttpClientHandlers that are platform specific.
+    /// See platform specific implementations for details on these issues.
+    /// </remarks>
+    internal class SimpleHttpClientFactory : IMsalHttpClientFactory
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/SimpleHttpClientFactory.cs:18]: We would still use this on Mac and NetStandard, but on Mac we only have PublicClient and we don't recommend pure NetStandard  implementation

File: src/client/Microsoft.Identity.Client/Http/HttpClientFactory.cs [hunk 3]
```diff
@@ -1,49 +0,0 @@
...
-    {
-        private readonly HttpClient _httpClient;
-
-        // The HttpClient is a singleton per ClientApplication so that we don't have a process wide singleton.
-        public const long MaxResponseContentBufferSizeInBytes = 1024 * 1024;
-
-        public HttpClientFactory()
-        {
-#if iOS
-            // See https://aka.ms/msal-net-httpclient for details
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Http/HttpClientFactory.cs:23]: >// See https://aka.ms/msal-net-httpclient for details [](start = 12, length = 53)

want to keep this one around in the config or in the httpmanager?

File: tests/devapps/NetCoreTestApp/Program.cs [hunk 4]
```diff
@@ -204,6 +222,25 @@ 1. IWA
...
+            return cca;
+        }
+
+        // Simple confidential client app with access to https://graph.microsoft.com/.default
+        private static readonly string s_clientIdForConfidentialApp =
+            Environment.GetEnvironmentVariable("LAB_APP_CLIENT_ID");
+
+        // App secret for app above 
+        private static readonly string s_confidentialClientSecret =
+            Environment.GetEnvironmentVariable("LAB_APP_CLIENT_SECRET");
```
Comment (henrik-me) [inline tests/devapps/NetCoreTestApp/Program.cs:241]: nit: top of file?

File: src/client/Microsoft.Identity.Client/Platforms/net45/Http/DnsSensitiveClientHandler.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+using System.Collections.Concurrent;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Http;
+
+namespace Microsoft.Identity.Client.Platforms.net45.Http
+{
+    internal class DnsSensitiveClientHandler : DelegatingHandler
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/net45/Http/DnsSensitiveClientHandler.cs:14]: >DnsSensitiveClientHandler [](start = 19, length = 25)

isn't this also needed for netcore?

Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/Platforms/net45/Http/DnsSensitiveClientHandler.cs:14]: ah, you are setting it just not using this helper class...

---
In reply to: [446375812](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1906#discussion_r446375812) [](ancestors = 446375812)

File: tests/devapps/NetFxConsoleTestApp/Program.cs [hunk 6]
```diff
@@ -294,11 +293,22 @@ 1. IWA
...
+                                    new[] { "https://graph.microsoft.com/.default" })
+                                    .WithForceRefresh(true)
+                                    .ExecuteAsync()
+                                    .ConfigureAwait(false);
+
+                                await Task.Delay(500).ConfigureAwait(false);
+                                Console.WriteLine("Got a token");
+                            }
+
+                            Console.WriteLine("Finished");
```
Comment (henrik-me) [inline tests/devapps/NetFxConsoleTestApp/Program.cs:310]: >Console.WriteLine("Finished"); [](start = 28, length = 30)

is there something we can measure here ensuring we only use 1 socket?

**Comment by jennyf19** [comment]
@bgavrilMS i like how you separate this by platform. we're doing better at separating things.

File: src/client/Microsoft.Identity.Client/Platforms/net45/Http/DnsSensitiveClientHandler.cs (hunk 5 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Platforms/net45/Http/DnsSensitiveClientHandler.cs:14]: On .net core they've left out `ServicePointManager.FindServicePoint`. Luckily they introduced a mechanism in .net core 2.1 and MSAL targets exactly .net core 2.1 ... That's why the .net core implementation is much simpler.

**Pull request merged**